### PR TITLE
Add blog article about volume mode conversion going GA

### PR DESCRIPTION
**Pull request by RaunakShah**
~Placeholder~ PR for volume mode conversion stable blog post
KEP - https://github.com/kubernetes/enhancements/issues/3141

### CODE REVIEW

**Comment by fsmunoz** [comment]
Hi @RaunakShah , Release Lead Shadow (assisting Docs) here, could you target this to `dev-1.30` instead of `main`? Thanks!

**Comment by a-mccarthy** [comment]
@RaunakShah 👋 just following up on your feature blog post for 1.30. The Ready for Review deadline is coming up next week on March 22, 2024. thanks!

please also note the comment above related to the docs ready for review deadline, https://github.com/kubernetes/website/pull/45262#issuecomment-1979194823

**Comment by kcmartin** [comment]
Hi @RaunakShah ! 1.30 Comms Lead here! 

I apologize for any confusion between the Docs and Comms teams, but this blog post should target `main` -- the reason for this is that Feature Blogs are published after the 1.30 release happens.

Please let us know if you have any questions

**Comment by sftim** [comment]
I'll mark this as WIP in a way that Prow honors
/retitle [WIP] Add blog article about volume mode conversion going GA

**Comment by a-mccarthy** [comment]
Hi @RaunakShah Hey there! This is a reminder about the upcoming Ready for Review deadline, this Friday, March 22, 2024. You'll need to have added the blog post to this PR by then so that the Blog Team and the Release Comms team can review your work in time for publication. Let me know if you have any questions 🙏

**Comment by RaunakShah** [comment]
Hi @a-mccarthy this PR is ready to be reviewed! 
cc @xing-yang

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+---
+layout: blog
+title: "Kubernetes 1.30: Preventing unauthorised volume mode conversion moves to GA"
+date: 2024-03-22
```
Comment (a-mccarthy) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:4]: The planned publication date for this blog is `2024-04-30`

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+This presents a security gap that allows malicious users to potentially exploit an
+as-yet-unknown vulnerability in the host operating system.
+
+There is a valid use case to allow some users to perform such conversions. Typically, storage backup
+vendors convert the volume mode during the course of a backup operation for greater efficiency of
+operations. This prevents Kubernetes from blocking the operation completely and presents a challenge
+in distinguishing trusted users from malicious ones.
+
+### Preventing unauthorised users from converting the volume mode
```
Comment (sftim) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:39]:
```suggestion
## Preventing unauthorised users from converting the volume mode
```

**Review by sftim — COMMENTED** [review]
(the heading levels need a tweak)

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+With the release of Kubernetes 1.30, the feature to prevent the modification of the volume mode
+of a [`PersistentVolumeClaim`](/docs/concepts/storage/persistent-volumes/) that was created from
+an existing VolumeSnapshot in a Kubernetes cluster, has moved to GA!
+
+
+### The problem
+
+The [Volume Mode](/docs/concepts/storage/persistent-volumes/#volume-mode) of a `PersistentVolumeClaim` 
+whether the underlying volume on the storage device is formatted into a filesystem or presented as
```
Comment (xing-yang) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:18]: The Volume Mode of a `PersistentVolumeClaim` refers to whether ...

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+for more details on how to create a `PVC` from an existing `VolumeSnapshot` in a Kubernetes cluster.
+
+When leveraging the above capability, there is no logic that validates whether the mode of the
+original volume, whose snapshot was taken, matches the mode of the newly created volume.
+
+This presents a security gap that allows malicious users to potentially exploit an
+as-yet-unknown vulnerability in the host operating system.
+
+There is a valid use case to allow some users to perform such conversions. Typically, storage backup
+vendors convert the volume mode during the course of a backup operation for greater efficiency of
```
Comment (xing-yang) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:35]: ... to retrieve changed blocks  for greater efficiency ...

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+1. Identify the `VolumeSnapshot` that is to be used as the data source for a newly
+   created `PVC` in the given namespace.
+2. Identify the `VolumeSnapshotContent` bound to the above `VolumeSnapshot`.
+
+  ```shell
+  kubectl describe volumesnapshot -n <namespace> <name>
+  ```
+
+3. Add the annotation [`snapshot.storage.kubernetes.io/allow-volume-mode-change: "true"`](/docs/reference/labels-annotations-taints/#snapshot-storage-kubernetes-io-allowvolumemodechange)
```
Comment (xing-yang) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:58]: following manifest fragment -> the following manifest fragment

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+Users should keep this in mind before they attempt to add the annotation
+to any `VolumeSnapshotContent`.
+
+
+### What's next
+
+To determine which CSI external sidecar versions support this feature, please head
+over to the [CSI docs page](https://kubernetes-csi.github.io/docs/).
+For any queries or issues, join [Kubernetes on Slack](https://slack.k8s.io/) and
+create a thread in the #sig-storage channel. Alternately, create an issue in the
```
Comment (xing-yang) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:108]: Can you add the #csi slack channel?

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+There is a valid use case to allow some users to perform such conversions. Typically, storage backup
+vendors convert the volume mode during the course of a backup operation, to retrieve changed blocks 
+for greater efficiency of operations. This prevents Kubernetes from blocking the operation completely
+and presents a challenge in distinguishing trusted users from malicious ones.
+
+## Preventing unauthorised users from converting the volume mode
+
+In this context, an authorised user is one who has access rights to perform `Update`
+or `Patch` operations on `VolumeSnapshotContents`, which is a cluster-level resource.  
+It is upto the cluster administrator to provide these rights only to trusted users
```
Comment (stmcginnis) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:43]:
```suggestion
It is up to the cluster administrator to provide these rights only to trusted users
```

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+or applications, like backup vendors.
+Users apart from such unauthorised ones will never be allowed to modify the volume mode
```
Comment (stmcginnis) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:45]: Do you mean "from such _authorised_ ones" here, referring to the authorized backup vendors use case?

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md (hunk 1 above)
Comment (RaunakShah) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md]: Fixed

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md (hunk 4 above)
Comment (RaunakShah) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md]: Done

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md (hunk 6 above)
Comment (RaunakShah) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md]: Done

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md (hunk 5 above)
Comment (RaunakShah) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:58]: Done

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md (hunk 8 above)
Comment (RaunakShah) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md]: Ooh good catch, fixed!

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+1. Identify the `VolumeSnapshot` that is to be used as the data source for a newly
+   created `PVC` in the given namespace.
+2. Identify the `VolumeSnapshotContent` bound to the above `VolumeSnapshot`.
+
+  ```shell
+  kubectl describe volumesnapshot -n <namespace> <name>
+  ```
+
+3. Add the annotation [`snapshot.storage.kubernetes.io/allow-volume-mode-change: "true"`](/docs/reference/labels-annotations-taints/#snapshot-storage-kubernetes-io-allowvolumemodechange)
+   to the above `VolumeSnapshotContent`. The `VolumeSnapshotContent` annotation must look like the following manifest fragment:
```
Comment (sftim) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:59]: (nit)
```suggestion
   to the above VolumeSnapshotContent. The VolumeSnapshotContent annotations must include one similar to the the following manifest fragment:
```

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+and presents a challenge in distinguishing trusted users from malicious ones.
+
+## Preventing unauthorised users from converting the volume mode
+
+In this context, an authorised user is one who has access rights to perform `Update`
+or `Patch` operations on `VolumeSnapshotContents`, which is a cluster-level resource.  
+It is up to the cluster administrator to provide these rights only to trusted users
+or applications, like backup vendors.
+Users apart from such authorised ones will never be allowed to modify the volume mode
+of a `PVC` when it is being created from a `VolumeSnapshot`.
```
Comment (sftim) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:46]: Please update the article to match our style guide for API kinds (no backticks), and to write API verbs in lowercase bold:

```suggestion
In this context, an authorised user is one who has access rights to perform **update**
or **patch operations on VolumeSnapshotContents, which is a cluster-level resource.  
It is up to the cluster administrator to provide these rights only to trusted users
or applications, like backup vendors.
Users apart from such authorised ones will never be allowed to modify the volume mode
of a PVC when it is being created from a VolumeSnapshot.
```

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+depending on the mode of the volume from which this snapshot was taken.
+
+An example is shown below:
+
+   ```yaml
+   apiVersion: snapshot.storage.k8s.io/v1
+   kind: VolumeSnapshotContent
+   metadata:
+     annotations:
+     - snapshot.storage.kubernetes.io/allowVolumeModeChange: "true"
```
Comment (sftim) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:80]: We should register this annotation before we release code for it, and we should try hard not to publish an announcement before the docs are all done.

What needs doing: update https://kubernetes.io/docs/reference/labels-annotations-taints/ on the _dev-1.30_ branch, to include a mention of this annotation.

**Review by sftim — COMMENTED** [review]
Thanks

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md (hunk 10 above)
Comment (RaunakShah) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md]: Thanks for the info, I've made that change here and elsewhere in the article.

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md (hunk 11 above)
Comment (RaunakShah) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md]: There was a mistake in the manifest fragment. The updated annotation already exists in the docs - https://kubernetes.io/docs/reference/labels-annotations-taints/#snapshot-storage-kubernetes-io-allow-volume-mode-change
I've made the change here

Comment (sftim) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md]: Ah, cool. That explains it - I did a string search for the annotation key as I saw it.

**Review by sftim — COMMENTED** [review]
LGTM (I didn't give it a proper look)

Please squash it to 1 commit, this is likely good to publish.

**Comment by RaunakShah** [comment]
All commits have been squashed into one!

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+Repeat steps 1 to 3 for all VolumeSnapshotContents whose volume mode needs to be
+converted during a backup or restore operation. This can be done either via software
+with credentials of an authorised user or manually by the authorised user(s).
+
+If the annotation shown above is present on a VolumeSnapshotContent object,
+Kubernetes will not prevent the volume mode from being converted.
+Users should keep this in mind before they attempt to add the annotation
+to any VolumeSnapshotContent.
+
+
```
Comment (xing-yang) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:102]: We should add an Action Required Section to highlight what's in the "Urgent Upgrade Notes" sections of external-provisioner 4.0.0 and external-snapshotter v7.0.0:

https://github.com/kubernetes-csi/external-provisioner/releases/tag/v4.0.0
https://github.com/kubernetes-csi/external-snapshotter/releases/tag/v7.0.0

**Comment by sftim** [comment]
/hold

:stop_sign: Do not merge until Kubernetes v1.30 is released :stop_sign:

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md (hunk 12 above)
Comment (sftim) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:102]: SGTM. Do you want that added to the main release announcement as well?

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md (hunk 8 above)
Comment (sftim) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md]: Although, US spelling, I think we prefer unauthorized

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md (hunk 3 above)
Comment (sftim) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md]: Even better: no backticks

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md (hunk 12 above)
Comment (a-mccarthy) [reply content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:102]: @sftim and @xing-yang I've added a note about this to the release blog. @RaunakShah are you able to add in a note about this?

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+Kubernetes will not prevent the volume mode from being converted.
+Users should keep this in mind before they attempt to add the annotation
+to any VolumeSnapshotContent.
+
+## Action Required
+
+The `prevent-volume-mode-conversion` feature flag is enabled by default in the 
+external-provisioner `v4.0.0` and snapshot-controller `v7.0.0` CSI sidecar 
+containers. Volume mode change will be rejected when creating a PVC from
+a VolumeSnapshot unless the steps described above have been performed.
```
Comment (xing-yang) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:107]: One more comment: can you just change to "external-provisioner `v4.0.0` and external-snapshotter `v7.0.0`?  The flag is available in snapshot-controller but not really available in csi-snapshotter sidecar.

File: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+Repeat steps 1 to 3 for all VolumeSnapshotContents whose volume mode needs to be
+converted during a backup or restore operation. This can be done either via software
+with credentials of an authorised user or manually by the authorised user(s).
+
+If the annotation shown above is present on a VolumeSnapshotContent object,
+Kubernetes will not prevent the volume mode from being converted.
+Users should keep this in mind before they attempt to add the annotation
+to any VolumeSnapshotContent.
+
+## Action Required
```
Comment (sftim) [inline content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md:102]: (nit)
```suggestion
## Action required
```

**Review by sftim — COMMENTED** [review]
/lgtm
/approve

Do not unhold until v1.30 is released.

**Comment by RaunakShah** [comment]
> Even better: fix https://github.com/kubernetes/website/pull/45262#discussion_r1555437056, squash to 1 commit.

Done!

**Review by sftim — COMMENTED** [review]
/hold cancel

**Pull request merged**